File: apps/frontend/src/stores/llm.ts
/**
 * LLM Store
 *
 * Zustand store for LLM chat and knowledge base functionality.
 */

import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import api from '@/services/api';

// Types
export type ChatRole = 'system' | 'user' | 'assistant' | 'function';

export interface ChatMessage {
  id: string;
  role: ChatRole;
  content: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  created_at: string;
  completed_at?: string;
  token_count: number;
  error?: string;
  metadata?: Record<string, any>;
}

export interface ChatSession {
  session_id: string;
  user_id: number;
  model: string;
  system_prompt?: string;
  parameters: Record<string, any>;
  message_count: number;
  created_at: string;
  updated_at: string;
  title?: string;
}

export interface LLMModel {
  id: string;
  type: string;
  context_window: number;
  supports_streaming: boolean;
  supports_function_calling: boolean;
}

export interface KnowledgeBase {
  id: string;
  name: string;
  description?: string;
  user_id: number;
  embedding_model: string;
  chunk_size: number;
  chunk_overlap: number;
  retrieval_top_k: number;
  document_count: number;
  created_at: string;
  updated_at: string;
}

export interface Document {
  id: string;
  kb_id: string;
  title: string;
  content: string;
  source_uri?: string;
  source_type: string;
  mime_type?: string;
  file_size?: number;
  chunk_count: number;
  status: string;
  created_at: string;
}

export interface SearchResult {
  chunk_id: string;
  document_id: string;
  content: string;
  score: number;
  metadata?: Record<string, any>;
}

export interface RAGAnswer {
  answer: string;
  sources: Array<{
    chunk_id: string;
    document_id: string;
    score: number;
  }>;
  context_used: number;
}

interface LLMState {
  // Chat state
  sessions: ChatSession[];
  currentSession: ChatSession | null;
  currentMessages: ChatMessage[];
  availableModels: LLMModel[];
  streaming: boolean;
  loading: boolean;
  error: string | null;

  // Knowledge base state
  knowledgeBases: KnowledgeBase[];
  currentKB: KnowledgeBase | null;
  kbDocuments: Document[];
  searchResults: SearchResult[];

  // Chat actions
  fetchSessions: () => Promise<void>;
  createSession: (data: {
    model: string;
    system_prompt?: string;
    parameters?: Record<string, any>;
  }) => Promise<void>;
  fetchSession: (sessionId: string) => Promise<void>;
  deleteSession: (sessionId: string) => Promise<void>;
  clearHistory: (sessionId: string) => Promise<void>;
  sendMessage: (sessionId: string, message: string, stream?: boolean) => Promise<void>;
  switchModel: (sessionId: string, newModel: string) => Promise<void>;
  updateParameters: (sessionId: string, parameters: Record<string, any>) => Promise<void>;
  fetchModels: () => Promise<void>;

  // Knowledge base actions
  fetchKnowledgeBases: () => Promise<void>;
  createKnowledgeBase: (data: {
    name: string;
    description?: string;
    embedding_model?: string;
    chunk_size?: number;
    chunk_overlap?: number;
    retrieval_top_k?: number;
  }) => Promise<void>;
  deleteKnowledgeBase: (kbId: string) => Promise<void>;
  addDocument: (kbId: string, data: {
    title: string;
    content: string;
    source_uri?: string;
    source_type?: string;
    chunk_strategy?: string;
  }) => Promise<void>;
  deleteDocument: (kbId: string, documentId: string) => Promise<void>;
  searchKnowledgeBase: (kbId: string, query: string, topK?: number) => Promise<void>;
  askQuestion: (kbId: string, question: string, sessionId?: string) => Promise<RAGAnswer>;

  // State management
  setCurrentSession: (session: ChatSession | null) => void;
  setCurrentKB: (kb: KnowledgeBase | null) => void;
  clearError: () => void;
}

export const useLLMStore = create<LLMState>()(
  persist(
    (set, get) => ({
    // Initial state
    sessions: [],
    currentSession: null,
    currentMessages: [],
    availableModels: [],
    streaming: false,
    loading: false,
    error: null,
    knowledgeBases: [],
    currentKB: null,
    kbDocuments: [],
    searchResults: [],

    // Chat actions
    fetchSessions: async () => {
      set({ loading: true, error: null });
      try {
        const response = await api.get('/llm/chat/sessions');
        set({ sessions: response.data, loading: false });
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to fetch sessions',
          loading: false,
        });
        throw error;
      }
    },

    createSession: async (data) => {
      set({ loading: true, error: null });
      try {
        const response = await api.post('/llm/chat/sessions', data);
        const session = response.data;
        set((state) => ({
          sessions: [session, ...state.sessions],
          currentSession: session,
          currentMessages: [],
          loading: false,
        }));
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to create session',
          loading: false,
        });
        throw error;
      }
    },

    fetchSession: async (sessionId: string) => {
      set({ loading: true, error: null });
      try {
        const response = await api.get(`/llm/chat/sessions/${sessionId}`);
        set({
          currentSession: response.data,
          loading: false,
        });
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to fetch session',
          loading: false,
        });
        throw error;
      }
    },

    deleteSession: async (sessionId: string) => {
      set({ loading: true, error: null });
      try {
        await api.delete(`/llm/chat/sessions/${sessionId}`);
        set((state) => ({
          sessions: state.sessions.filter((s) => s.session_id !== sessionId),
          currentSession: state.currentSession?.session_id === sessionId ? null : state.currentSession,
          loading: false,
        }));
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to delete session',
          loading: false,
        });
        throw error;
      }
    },

    clearHistory: async (sessionId: string) => {
      set({ loading: true, error: null });
      try {
        await api.post(`/llm/chat/sessions/${sessionId}/clear`);
        set((state) => ({
          currentMessages: [],
          loading: false,
        }));
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to clear history',
          loading: false,
        });
        throw error;
      }
    },

    sendMessage: async (sessionId: string, message: string, stream = false) => {
      // Add user message optimistically
      const userMessage: ChatMessage = {
        id: `msg-${Date.now()}-user`,
        role: 'user',
        content: message,
        status: 'completed',
        created_at: new Date().toISOString(),
        token_count: 0,
      };

      set((state) => ({
        currentMessages: [...state.currentMessages, userMessage],
      }));

      // Create placeholder for assistant response
      const assistantMessage: ChatMessage = {
        id: `msg-${Date.now()}-assistant`,
        role: 'assistant',
        content: '',
        status: 'processing',
        created_at: new Date().toISOString(),
        token_count: 0,
      };

      set((state) => ({
        currentMessages: [...state.currentMessages, assistantMessage],
        loading: true,
      }));

      try {
        const response = await api.post(`/llm/chat/sessions/${sessionId}/message`, {
          message,
          stream,
        });

        set((state) => ({
          currentMessages: state.currentMessages.map((msg) =>
            msg.id === assistantMessage.id
              ? { ...msg, ...response.data, status: 'completed' as const }
              : msg
          ),
          loading: false,
        }));
      } catch (error: any) {
        set((state) => ({
          currentMessages: state.currentMessages.map((msg) =>
            msg.id === assistantMessage.id
              ? { ...msg, status: 'failed' as const, error: 'Failed to get response' }
              : msg
          ),
          loading: false,
          error: error.response?.data?.detail || 'Failed to send message',
        }));
        throw error;
      }
    },

    switchModel: async (sessionId: string, newModel: string) => {
      try {
        const response = await api.put(`/llm/chat/sessions/${sessionId}/model`, null, {
          params: { new_model: newModel },
        });
        set({ currentSession: response.data });
      } catch (error: any) {
        set({ error: error.response?.data?.detail || 'Failed to switch model' });
        throw error;
      }
    },

    updateParameters: async (sessionId: string, parameters: Record<string, any>) => {
      try {
        const response = await api.put(`/llm/chat/sessions/${sessionId}/parameters`, parameters);
        set({ currentSession: response.data });
      } catch (error: any) {
        set({ error: error.response?.data?.detail || 'Failed to update parameters' });
        throw error;
      }
    },

    fetchModels: async () => {
      try {
        const response = await api.get('/llm/chat/models');
        set({ availableModels: response.data });
      } catch (error: any) {
        set({ error: error.response?.data?.detail || 'Failed to fetch models' });
      }
    },

    // Knowledge base actions
    fetchKnowledgeBases: async () => {
      set({ loading: true, error: null });
      try {
        const response = await api.get('/llm/knowledge-bases');
        set({ knowledgeBases: response.data, loading: false });
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to fetch knowledge bases',
          loading: false,
        });
        throw error;
      }
    },

    createKnowledgeBase: async (data) => {
      set({ loading: true, error: null });
      try {
        const response = await api.post('/llm/knowledge-bases', data);
        set((state) => ({
          knowledgeBases: [...state.knowledgeBases, response.data],
          loading: false,
        }));
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to create knowledge base',
          loading: false,
        });
        throw error;
      }
    },

    deleteKnowledgeBase: async (kbId: string) => {
      set({ loading: true, error: null });
      try {
        await api.delete(`/llm/knowledge-bases/${kbId}`);
        set((state) => ({
          knowledgeBases: state.knowledgeBases.filter((kb) => kb.id !== kbId),
          loading: false,
        }));
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to delete knowledge base',
          loading: false,
        });
        throw error;
      }
    },

    addDocument: async (kbId: string, data) => {
      set({ loading: true, error: null });
      try {
        const response = await api.post(`/llm/knowledge-bases/${kbId}/documents`, data);
        set((state) => ({
          loading: false,
        }));
        return response.data;
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to add document',
          loading: false,
        });
        throw error;
      }
    },

    deleteDocument: async (kbId: string, documentId: string) => {
      set({ loading: true, error: null });
      try {
        await api.delete(`/llm/knowledge-bases/${kbId}/documents/${documentId}`);
        set({ loading: false });
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to delete document',
          loading: false,
        });
        throw error;
      }
    },

    searchKnowledgeBase: async (kbId: string, query: string, topK?: number) => {
      set({ loading: true, error: null });
      try {
        const response = await api.post(`/llm/knowledge-bases/${kbId}/search`, {
          query,
          top_k: topK,
        });
        set({ searchResults: response.data, loading: false });
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to search',
          loading: false,
        });
        throw error;
      }
    },

    askQuestion: async (kbId: string, question: string, sessionId?: string) => {
      set({ loading: true, error: null });
      try {
        const response = await api.post(`/llm/knowledge-bases/${kbId}/answer`, {
          question,
          session_id: sessionId,
        });
        set({ loading: false });
        return response.data;
      } catch (error: any) {
        set({
          error: error.response?.data?.detail || 'Failed to get answer',
          loading: false,
        });
        throw error;
      }
    },

    // State management
    setCurrentSession: (session) => set({ currentSession: session }),
    setCurrentKB: (kb) => set({ currentKB: kb }),
    clearError: () => set({ error: null }),
  }),
  {
    name: 'llm-store',
    partialize: (state) => ({
      sessions: state.sessions,
      currentSession: state.currentSession,
      knowledgeBases: state.knowledgeBases,
      currentKB: state.currentKB,
    }),
  }
));

// Selectors
export const useChatSessions = () => useLLMStore((state) => state.sessions);
export const useCurrentChatSession = () => useLLMStore((state) => state.currentSession);
export const useChatMessages = () => useLLMStore((state) => state.currentMessages);
export const useLLMLoading = () => useLLMStore((state) => state.loading);
export const useLLMError = () => useLLMStore((state) => state.error);
export const useKnowledgeBases = () => useLLMStore((state) => state.knowledgeBases);
export const useSearchResults = () => useLLMStore((state) => state.searchResults);

File: apps/backend/app/api/v1/llm.py
"""
LLM API Endpoints

REST API for LLM chat, knowledge base, and inference management.
"""

from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from pydantic import BaseModel, Field

from app.api.deps import get_current_user
from app.models.user import User
from app.services.llm import (
    llm_chat_service,
    rag_service,
    ChatRole,
    MessageStatus,
    ChunkStrategy,
)

router = APIRouter(prefix="/llm", tags=["llm"])

# Request/Response Models


class CreateSessionRequest(BaseModel):
    """Request to create a chat session"""

    model: str = Field(..., description="Model to use for chat")
    system_prompt: Optional[str] = Field(
        None, description="System prompt for the session"
    )
    parameters: Optional[dict] = Field(
        None, description="Chat parameters (temperature, max_tokens, etc.)"
    )


class ChatRequest(BaseModel):
    """Request to send a chat message"""

    message: str = Field(..., min_length=1, max_length=10000, description="User message")
    stream: bool = Field(False, description="Stream the response")
    parameters: Optional[dict] = Field(None, description="Override parameters")


class ChatResponse(BaseModel):
    """Chat message response"""

    id: str
    role: str
    content: str
    status: str
    created_at: str
    completed_at: Optional[str]
    token_count: int


class CreateKnowledgeBaseRequest(BaseModel):
    """Request to create a knowledge base"""

    name: str = Field(..., min_length=1, max_length=256, description="KB name")
    description: Optional[str] = Field(None, description="KB description")
    embedding_model: str = Field("bge-large-zh", description="Embedding model")
    chunk_size: int = Field(500, ge=100, le=2000, description="Chunk size")
    chunk_overlap: int = Field(50, ge=0, le=500, description="Chunk overlap")
    retrieval_top_k: int = Field(5, ge=1, le=20, description="Top K retrieval")


class AddDocumentRequest(BaseModel):
    """Request to add a document to knowledge base"""

    title: str = Field(..., min_length=1, max_length=512, description="Document title")
    content: str = Field(..., min_length=1, description="Document content")
    source_uri: Optional[str] = Field(None, description="Source URI")
    source_type: str = Field("upload", description="Source type")
    mime_type: Optional[str] = Field(None, description="MIME type")
    chunk_strategy: str = Field("fixed_size", description="Chunking strategy")


class SearchRequest(BaseModel):
    """Request to search knowledge base"""

    query: str = Field(..., min_length=1, description="Search query")
    top_k: Optional[int] = Field(None, ge=1, le=20, description="Number of results")
    document_id: Optional[str] = Field(None, description="Filter by document")


class RAGQuestionRequest(BaseModel):
    """Request to ask a question with RAG"""

    question: str = Field(..., min_length=1, description="Question")
    session_id: Optional[str] = Field(None, description="Chat session ID")


# Chat Endpoints


@router.post("/chat/sessions")
async def create_chat_session(
    request: CreateSessionRequest,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Create a new chat session"""
    session = llm_chat_service.create_session(
        user_id=current_user.id,
        model=request.model,
        system_prompt=request.system_prompt,
        parameters=request.parameters,
    )
    return session.to_dict()


@router.get("/chat/sessions")
async def list_chat_sessions(
    limit: int = Query(50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
) -> List[dict]:
    """List chat sessions for current user"""
    sessions = llm_chat_service.list_sessions(current_user.id, limit)
    return [s.to_dict() for s in sessions]


@router.get("/chat/sessions/{session_id}")
async def get_chat_session(
    session_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Get chat session details"""
    session = llm_chat_service.get_session(session_id)
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session {session_id} not found"
        )
    return session.to_dict()


@router.delete("/chat/sessions/{session_id}")
async def delete_chat_session(
    session_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Delete a chat session"""
    success = llm_chat_service.delete_session(session_id, current_user.id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session {session_id} not found"
        )
    return {"deleted": True}


@router.post("/chat/sessions/{session_id}/clear")
async def clear_chat_history(
    session_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Clear message history for a session"""
    success = llm_chat_service.clear_history(session_id, current_user.id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session {session_id} not found"
        )
    return {"cleared": True}


@router.post("/chat/sessions/{session_id}/message")
async def send_chat_message(
    session_id: str,
    request: ChatRequest,
    current_user: User = Depends(get_current_user),
) -> ChatResponse:
    """Send a message and get a response"""
    result = await llm_chat_service.chat(
        session_id=session_id,
        user_message=request.message,
        user_id=current_user.id,
        stream=request.stream,
        parameters=request.parameters,
    )

    if isinstance(result, str):
        # Streaming would be handled via Server-Sent Events
        raise HTTPException(
            status_code=status.HTTP_501_NOT_IMPLEMENTED,
            detail="Streaming not yet implemented via REST"
        )

    return ChatResponse(**result.to_dict())


@router.put("/chat/sessions/{session_id}/model")
async def switch_session_model(
    session_id: str,
    new_model: str = Query(..., description="New model to use"),
    current_user: User = Depends(get_current_user),
) -> dict:
    """Switch the model for a session"""
    try:
        session = llm_chat_service.switch_model(session_id, new_model, current_user.id)
        return session.to_dict()
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.put("/chat/sessions/{session_id}/parameters")
async def update_session_parameters(
    session_id: str,
    parameters: dict,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Update session parameters"""
    session = llm_chat_service.update_parameters(
        session_id, parameters, current_user.id
    )
    return session.to_dict()


@router.get("/chat/models")
async def list_available_models(
    supports_streaming: Optional[bool] = Query(None),
    current_user: User = Depends(get_current_user),
) -> List[dict]:
    """List available LLM models"""
    models = llm_chat_service.router.list_models(
        supports_streaming=supports_streaming
    )
    return models


# Knowledge Base Endpoints


@router.post("/knowledge-bases")
async def create_knowledge_base(
    request: CreateKnowledgeBaseRequest,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Create a new knowledge base"""
    kb = rag_service.create_knowledge_base(
        name=request.name,
        description=request.description,
        user_id=current_user.id,
        embedding_model=request.embedding_model,
        chunk_size=request.chunk_size,
        chunk_overlap=request.chunk_overlap,
        retrieval_top_k=request.retrieval_top_k,
    )
    return kb.to_dict()


@router.get("/knowledge-bases")
async def list_knowledge_bases(
    current_user: User = Depends(get_current_user),
) -> List[dict]:
    """List knowledge bases for current user"""
    kbs = rag_service.list_knowledge_bases(current_user.id)
    return [kb.to_dict() for kb in kbs]


@router.get("/knowledge-bases/{kb_id}")
async def get_knowledge_base(
    kb_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Get knowledge base details"""
    kb = rag_service.get_knowledge_base(kb_id)
    if not kb:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Knowledge base {kb_id} not found"
        )
    return kb.to_dict()


@router.delete("/knowledge-bases/{kb_id}")
async def delete_knowledge_base(
    kb_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Delete a knowledge base"""
    success = rag_service.delete_knowledge_base(kb_id, current_user.id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Knowledge base {kb_id} not found"
        )
    return {"deleted": True}


@router.post("/knowledge-bases/{kb_id}/documents")
async def add_document(
    kb_id: str,
    request: AddDocumentRequest,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Add a document to a knowledge base"""
    try:
        document = await rag_service.add_document(
            kb_id=kb_id,
            title=request.title,
            content=request.content,
            source_uri=request.source_uri,
            source_type=request.source_type,
            mime_type=request.mime_type,
            chunk_strategy=ChunkStrategy(request.chunk_strategy),
        )
        return {
            "id": document.id,
            "kb_id": document.kb_id,
            "title": document.title,
            "chunk_count": len(document.chunks),
            "status": document.status.value,
            "created_at": document.created_at.isoformat(),
        }
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/knowledge-bases/{kb_id}/search")
async def search_knowledge_base(
    kb_id: str,
    request: SearchRequest,
    current_user: User = Depends(get_current_user),
) -> List[dict]:
    """Search a knowledge base"""
    try:
        results = await rag_service.search(
            kb_id=kb_id,
            query=request.query,
            top_k=request.top_k,
            document_id=request.document_id,
        )
        return results
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/knowledge-bases/{kb_id}/answer")
async def ask_question(
    kb_id: str,
    request: RAGQuestionRequest,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Ask a question using RAG"""
    try:
        result = await rag_service.answer(
            kb_id=kb_id,
            question=request.question,
            llm_service=llm_chat_service,
            session_id=request.session_id,
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.delete("/knowledge-bases/{kb_id}/documents/{document_id}")
async def delete_document(
    kb_id: str,
    document_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Delete a document from knowledge base"""
    success = rag_service.delete_document(kb_id, document_id, current_user.id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Document not found"
        )
    return {"deleted": True}


# Inference Service Endpoints


@router.post("/inference/completions")
async def create_completion(
    model: str = Query(..., description="Model to use"),
    prompt: str = Query(..., description="Prompt text"),
    max_tokens: int = Query(512, ge=1, le=4096, description="Max tokens"),
    temperature: float = Query(0.7, ge=0, le=2, description="Temperature"),
    stream: bool = Query(False, description="Stream response"),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    Create a text completion (OpenAI-compatible).

    This endpoint provides OpenAI-compatible completion API.
    """
    # In production, call actual inference service
    return {
        "id": "cmpl-" + str(hash(prompt) % 1000000),
        "object": "text_completion",
        "created": int(datetime.now().timestamp()),
        "model": model,
        "choices": [
            {
                "text": f"Completion for: {prompt[:50]}...",
                "index": 0,
                "logprobs": None,
                "finish_reason": "length",
            }
        ],
        "usage": {
            "prompt_tokens": len(prompt.split()),
            "completion_tokens": 20,
            "total_tokens": len(prompt.split()) + 20,
        },
    }


@router.post("/inference/chat/completions")
async def create_chat_completion(
    request: dict,
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    Create a chat completion (OpenAI-compatible).

    This endpoint provides OpenAI-compatible chat API.
    """
    model = request.get("model", "gpt-3.5-turbo")
    messages = request.get("messages", [])
    temperature = request.get("temperature", 0.7)
    max_tokens = request.get("max_tokens", 2048)
    stream = request.get("stream", False)

    # Extract last user message
    user_message = ""
    for msg in messages:
        if msg.get("role") == "user":
            user_message = msg.get("content", "")

    # In production, call actual inference service
    response_text = f"This is a simulated response to: {user_message[:100]}"

    return {
        "id": "chatcmpl-" + str(hash(user_message) % 1000000),
        "object": "chat.completion",
        "created": int(datetime.now().timestamp()),
        "model": model,
        "choices": [
            {
                "index": 0,
                "message": {
                    "role": "assistant",
                    "content": response_text,
                },
                "finish_reason": "stop",
            }
        ],
        "usage": {
            "prompt_tokens": sum(len(m.get("content", "").split()) for m in messages),
            "completion_tokens": len(response_text.split()),
            "total_tokens": sum(len(m.get("content", "").split()) for m in messages) + len(response_text.split()),
        },
    }


@router.get("/inference/endpoints")
async def list_inference_endpoints(
    current_user: User = Depends(get_current_user),
) -> List[dict]:
    """List available inference endpoints"""
    # Return deployed models from AIHub
    return [
        {
            "id": "chatglm3-6b",
            "name": "ChatGLM3-6B",
            "type": "llm",
            "endpoint": "/v1/chatglm3-6b",
            "status": "running",
            "created_at": "2024-01-01T00:00:00Z",
        },
        {
            "id": "qwen-14b",
            "name": "Qwen-14B",
            "type": "llm",
            "endpoint": "/v1/qwen-14b",
            "status": "running",
            "created_at": "2024-01-01T00:00:00Z",
        },
    ]


@router.get("/inference/endpoints/{endpoint_id}/metrics")
async def get_endpoint_metrics(
    endpoint_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """Get metrics for an inference endpoint"""
    return {
        "endpoint_id": endpoint_id,
        "requests_per_second": 5.2,
        "average_latency_ms": 150,
        "p95_latency_ms": 300,
        "p99_latency_ms": 500,
        "error_rate": 0.001,
        "token_throughput": 1250,
        "gpu_utilization": 0.75,
    }


from datetime import datetime
